sched/adaptdl_sched/allocator.py

```diff
@@ -38,27 +38,109 @@ class AdaptDLAllocator(object):
     def __init__(self, expander):
         self._core_api = kubernetes.client.CoreV1Api()
         self._objs_api = kubernetes.client.CustomObjectsApi()
+        self._custom_resource = ("adaptdl.petuum.com", "v1",
+                                 "", "adaptdljobs")
         self._cluster_expander = expander
         self._policy = PolluxPolicy()
+        # lock for the two corountines in run()
+        self.lock = asyncio.Lock()
 
     async def run(self):
+        # two functionality: (1) watch for new job and start if possible.
+        # (2) periodically optimize existing jobs
+        await asyncio.gather(
+            self._new_job_allocate(),
+            self._periodic_optimize()
+        )
+
+    async def _new_job_allocate(self):
+        async with kubernetes.watch.Watch() as watch:
+            while True:
+                async for event in watch.stream(
+                        self._objs_api.list_namespaced_custom_object,
+                        *self._custom_resource, timeout_seconds=60):
+                    self._new_job_helper(event)
+                   
+                   
+    async def _new_job_helper(self):
+        # if there is an arriving job
+        if event["type"] == "ADDED":
+            async with self.lock:
```

Comment:
Lock and if statement can go outside of this method, so this method can be more unit-testable in the future